openjdk/mmtkMutator.cpp

```diff
@@ -9,8 +9,9 @@ MMTkMutatorContext MMTkMutatorContext::bind(::Thread* current) {
 HeapWord* MMTkMutatorContext::alloc(size_t bytes, Allocator allocator) {
     // FIXME: Proper use of slow-path api
     HeapWord* o = (HeapWord*) ::alloc((MMTk_Mutator) this, bytes, HeapWordSize, 0, allocator);
-    // FIXME: Proper post alloc
-
+    // Post allococation. Currently we are only calling post_alloc in slowpath here.
+    // TODO: We also need to call them in the fastpath.
+    ::post_alloc((MMTk_Mutator) this, o, bytes, allocator);
```

Comment:
The OpenJDK binding was relying on the fact that the allocator (the bump pointer allocator) it previously used did not require calling `post_alloc()`. But For the new malloc allocator, we have to call `post_alloc()` to init the alloc bit. As we always use slowpath for the malloc allocator, so I think it is still correct.

What is missing is the fastpath for `post_alloc()` for bump pointer allocator. But that was and is not needed.

Reply:
We cannot put `post_alloc` to `MallocAllocator::alloc()`. In `alloc()`, we only have addresses, but in `post_alloc()`, we set alloc bits for object references.

Reply:
Why is it incorrect? Is it because for bump allocator, we now call `post_alloc` for slowpath, but do not call `post_alloc` for fastpath?

Before this PR:
|                  | Fastpath      | Slowpath   |   
|------------------|---------------|------------|
| Bump Allocator   | no post_alloc | no post_alloc |

In this PR:
|                  | Fastpath      | Slowpath   |
|------------------|---------------|------------|
| Bump Allocator   | no post_alloc | post_alloc |
| Malloc Allocator | -             | post_alloc |

Reply:
Umm, I see. It is not only about the allocator, but also the space in use.

Reply:
I think the problem is that for some policies, we may need to call `post_alloc` in bump allocator's fastpath. Always calling `post_alloc()` in the slowpath looks correct to me. If a policy requires nothing in `post_alloc`, it is fine that we just call it and do a noop. That does not sound like an issue.